### PR TITLE
MKLDNN Sum and Batch Normalizaton kernels

### PR DESCRIPTION
**Pull request by sreekanth-yalachigere**
**Sum Kernel**
```
core\providers\mkldnn\math\sum.cc
core\providers\mkldnn\math\sum.h
```

Modified base `class Sum_6` in `core\providers\cpu\math\element_wise_ops.h
`Removed final 
for mkldnn to fallback to cpu computation.

**Batch Norm Kernel**
```
core\providers\mkldnn\nn\batch_norm.cc
core\providers\mkldnn\nn\batch_norm.h
```

Modified base class `onnxruntime::BatchNorm` in file `core\providers\cpu\nn\batch_norm.h`
Removed _final_ and made _epsilon__ as protected member.

### CODE REVIEW

**Comment by sreekanth-yalachigere** [comment]
@pranavsharma how to kick off CI pipeline?

**Comment by sreekanth-yalachigere** [comment]
@jywu-msft I noticed it. I am having problems to sync master. 
getting the following message when I try to merge upstream/master

`merge: upsteam/master - not something we can merge
`
Can I close this create a new PR?

**Comment by jywu-msft** [comment]
did you fetch upstream to pull in the latest and then merge?

**Comment by sreekanth-yalachigere** [comment]
yes

**Comment by sreekanth-yalachigere** [comment]
@jywu-msft I now only see my changes. Can you please take a look?

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace mkl_dnn {
+
+template <typename T>
+class BatchNorm final : public onnxruntime::BatchNorm<T> {
+ public:
+   BatchNorm(const OpKernelInfo& info) : onnxruntime::BatchNorm<T>(info) {}
+
+  Status Compute(OpKernelContext* context) const override;
+
+ private:
```
Comment (yufenglee) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.h]: remove this

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+  mkldnn::memory::dims dst_dims_mkl(
+    Y->Shape().GetDims().begin(), Y->Shape().GetDims().end());
+
+  try {
+    BatchNormParams batchNorm_params(src_dims_mkl, scale_dims_mkl, 
+      b_dims_mkl, mean_dims_mkl, var_dims_mkl, dst_dims_mkl, 
+      onnxruntime::BatchNorm<T>::epsilon_, num_dimensions);
+    BatchNormPrimitive<T>* batchNorm_primitive = 
+      BatchNormPrimitivePool<T>::Get(batchNorm_params);
+    batchNorm_primitive->Compute(src_data, scale_data, b_data,
```
Comment (yufenglee) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: Is nullptr check needed here?

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.h (hunk 1 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/nn/batch_norm.h]: Will do.

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc (hunk 2 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: We are not checking for nullptr check in other primitives.  `BatchNormPrimitivePool<T>::Get `checks for batchNorm_primitive in the pool. If it is null, it will create new one and return.

I will add `ONNXRUNTIME_RETURN_IF_ERROR`. and also to Sum primitive.

Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: Added check.

`ONNXRUNTIME_RETURN_IF_NOT(batchNorm_primitive != nullptr);`

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+namespace {
+// Struct which encapsulates parameters for MKLDNN Sum primitives.
+struct SumParams {
+  std::vector<mkldnn::memory::dims> src_dims;
+  mkldnn::memory::dims& dst_dim;
+  int num_inputs;
+  int num_dimensions;
+
+  SumParams(std::vector<mkldnn::memory::dims> dims,
+    mkldnn::memory::dims& dst_dims, int &numinputs, int &dimensions)
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: any reason you're passing numinputs and dimensions by ref?

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+#include "core/providers/mkldnn/mkldnn_fwd.h"
+
+namespace onnxruntime {
+namespace mkl_dnn {
+
+namespace {
+// Struct which encapsulates parameters for MKLDNN Sum primitives.
+struct SumParams {
+  std::vector<mkldnn::memory::dims> src_dims;
+  mkldnn::memory::dims& dst_dim;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: why is dst_dim stored as a ref, while src_dims as value?

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+  ~SumPrimitive() = default;
+
+  void Compute(OpKernelContext* context, int& numinputs) {
+    const Tensor* X1 = context->Input<Tensor>(0);
+    Tensor* Y = context->Output(0, X1->Shape());
+    T* dst_data = Y->template MutableData<T>();
+
+    context_.dst_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(dst_data)));
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: why const_cast? dst_data is not even const to begin with.

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+      : cpu_engine_(GetEngine()) {
+    context_.stream.reset(new mkldnn::stream(mkldnn::stream::kind::eager));
+    if (context_.sum_pd == nullptr) {
+      Initialize(params);
+    }
+  }
+
+  ~SumPrimitive() = default;
+
+  void Compute(OpKernelContext* context, int& numinputs) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: ref for numinputs? is Compute updating numpinputs?

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+
+    std::unique_ptr<mkldnn::batch_normalization_forward::desc> batchnorm_fwd;
+    std::unique_ptr<mkldnn::batch_normalization_forward::primitive_desc> 
+      batchnorm_fwd_pd;
+
+    std::unique_ptr<mkldnn::stream> stream;
+    std::vector<mkldnn::primitive> net;
+
+    BatchNormContext()
+      : src_mem(nullptr),
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: unique_ptrs are nullptrs by default. there is no need to initialize them.

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+      const T* src_data = X->template Data<T>();
+      context_.srcs_memory[i].set_data_handle(
+        static_cast<void*>(const_cast<T*>(src_data)));
+    }
+
+    context_.stream->submit(context_.net);
+  }
+
+  std::unique_ptr<mkldnn::memory::desc> GetDstMemoryDesc() 
+    const { return context_.dst_md; }
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: const should be on the same line as the function

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+    std::vector<mkldnn::memory::primitive_desc> srcs_pd;
+    std::unique_ptr<mkldnn::memory::primitive_desc> src_mpd;
+    std::unique_ptr<mkldnn::memory::primitive_desc> dst_pd;
+    std::unique_ptr<mkldnn::sum::primitive_desc> sum_pd;
+
+    std::unique_ptr<mkldnn::stream> stream;
+    std::vector<mkldnn::primitive> net;
+
+    SumContext()
+      : src_md(nullptr),
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: unique_ptrs are null by default.

File: onnxruntime/core/providers/mkldnn/math/sum.cc (hunk 3 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/math/sum.cc]: yes. numinputs and dimensions should be pass by value. I will update it.

File: onnxruntime/core/providers/mkldnn/math/sum.cc (hunk 4 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/math/sum.cc]: src_dims is vector of dims. I will try changing it to ref.

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc (hunk 7 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: removed constructor in Sum and BatchNorm.

File: onnxruntime/core/providers/mkldnn/math/sum.cc (hunk 8 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/math/sum.cc]: done.

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+#include "core/providers/mkldnn/math/sum.h"
+#include "core/providers/mkldnn/mkldnn_fwd.h"
+
+namespace onnxruntime {
+namespace mkl_dnn {
+
+namespace {
+// Struct which encapsulates parameters for MKLDNN Sum primitives.
+struct SumParams {
+  std::vector<mkldnn::memory::dims>& src_dims;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: is there any reason these should be just ref and not const ref? same comment for dst_dim.

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+  mkldnn::memory::dims& dst_dim;
+  int num_inputs;
+  int num_dimensions;
+
+  SumParams(std::vector<mkldnn::memory::dims>& dims,
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: can these be passed by const-ref?

Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/math/sum.cc]: @pranavsharma src_dims can be const ref but we want dst_dims just refs. dst_dims will be modified after compute.

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+namespace {
+// Struct which encapsulates parameters for MKLDNN BatchNorm primitive.
+struct BatchNormParams {
+  mkldnn::memory::dims& src_dims;
+  mkldnn::memory::dims& scale_dims;
+  mkldnn::memory::dims& b_dims;
+  mkldnn::memory::dims& mean_dims;
+  mkldnn::memory::dims& var_dims;
+  mkldnn::memory::dims& dst_dims;
+  float epsilon;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: Differentiate between const-ref and just ref for those that will be modified.

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+  int num_dimensions;
+
+  BatchNormParams(mkldnn::memory::dims& src_dims_mkl,
+    mkldnn::memory::dims& scale_dims_mkl,
+    mkldnn::memory::dims& b_dims_mkl, mkldnn::memory::dims& mean_dims_mkl,
+    mkldnn::memory::dims& var_dims_mkl, mkldnn::memory::dims& dst_dims_mkl,
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: Differentiate between const-ref and just ref for those that will be modified.

Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: Okay. Will pass src_dims by const ref and dst_dims by ref.
